File: src/module/pairwise.cpp
/**
 * Museqa: Multiple Sequence Aligner using hybrid parallel computing.
 * @file Implementation for the heuristic's pairwise module.
 * @author Rodrigo Siqueira <rodriados@gmail.com>
 * @copyright 2020-present Rodrigo Siqueira
 */
#include <string>

#include "io.hpp"
#include "pipeline.hpp"
#include "exception.hpp"

#include "module/pairwise.cuh"

namespace museqa
{
    namespace module
    {
        using namespace pw = museqa::pairwise;

        /**
         * Execute the module's task when on a pipeline.
         * @param io The pipeline's IO service instance.
         * @return A conduit with the module's processed results.
         */
        auto pairwise::run(const io::service& io, const pairwise::pipe& pipe) const -> pairwise::pipe
        {
            auto algoname = io.get<std::string>(cli::pairwise, "default");
            auto tablename = io.get<std::string>(cli::scoring, "default");

            const auto conduit = pipeline::convert<pairwise::previous>(*pipe);

            auto table = pw::scoring_table::make(tablename);
            auto result = pw::run(*conduit.db, table, algoname);

            auto ptr = new pairwise::conduit {conduit.db, result};
            return pairwise::pipe {ptr};
        }

        /**
         * Checks whether command line arguments produce a valid module state.
         * @param io The pipeline's IO service instance.
         * @return Are the given command line arguments valid?
         */
        auto pairwise::check(const io::service& io) const -> bool
        {
            auto algoname = io.get<std::string>(cli::pairwise, "default");
            enforce(pw::algorithm::has(algoname), "unknown pairwise algorithm chosen: '%s'", algoname);

            auto tablename = io.get<std::string>(cli::scoring, "default");
            enforce(pw::scoring_table::has(tablename), "unknown scoring table chosen: '%s'", tablename);
            
            return true;
        }
    }
}

File: test/pairwise.py
#!/usr/bin/env python
# Museqa: Multiple Sequence Aligner using hybrid parallel computing.
# @file Unit tests for the project's pairwise module.
# @author Rodrigo Siqueira <rodriados@gmail.com>
# @copyright 2021-present Rodrigo Siqueira
from museqa.algorithm import pairwise as algorithm
from museqa.database import Database
from museqa import pairwise
import pytest

# Defines some fixture files to be processed by different test cases.
# @since 0.1.1
fixtures = [
    "fixtures/mock1.fasta"
,   "fixtures/mock2.fasta"
,   "fixtures/mock3.fasta"
]

# Creates a fixture for sequence databases.
# @param request The test context to run the fixture with.
# @since 0.1.1
@pytest.fixture(params = fixtures)
def database(request):
    return Database.load(request.param)

# Creates a fixture for different scoring tables.
# @param request The test context to run the fixture with.
# @since 0.1.1
@pytest.fixture(params = ['default', 'blosum62', 'pam250'])
def table(request):
    return request.param

# Compares whether two distance matrices and asserts whether they're equal.
# @param expected The expected distance matrix to be asserted.
# @param produced The produced distance matrix to be asserted.
def assertTablesAreEqual(expected, produced):
    for i in range(expected.count):
        for j in range(produced.count):
            assert expected[i, j] == produced[i, j]

# Runs the pairwise module with different algorithms and compare their results.
# @param db The database to test the algorithms with.
# @param reference The reference function to compare the generated result with.
# @param target The target algorithm function to test the module with.
# @param extra Extra arguments to test the selected algorithm with.
def assertAlgorithmExecution(db, reference, target, **extra):
    expected = pairwise.run(db, algorithm = reference, **extra)
    produced = pairwise.run(db, algorithm = target, **extra)
    assert expected.count == produced.count
    
    assertTablesAreEqual(expected, produced)

# Tests whether the sequential needleman algorithm produces the expected matrix.
# @param database The database to test the algorithm with.
# @param table The scoring table to run the algorothm with.
# @since 0.1.1
def testSequentialNeedleman(database, table):
    assertAlgorithmExecution(database, 'sequential', algorithm.needleman, table = table)

# Tests whether the hybrid needleman algorithm produces the expected matrix.
# @param database The database to test the algorithm with.
# @param table The scoring table to run the algorothm with.
# @since 0.1.1
@pytest.mark.skip(reason = "a CUDA device may not be available")
def testHybridNeedleman(database, table):
    assertAlgorithmExecution(database, 'hybrid', algorithm.needleman, table = table)
